### ISSUE TITLE
Difficulty including Link

### ISSUE THREAD

**Issue by brunchboy** [Ableton/link#15]
I am trying to follow the directions on how to build a project which embeds Link, and am running into some trouble. It is quite likely due to me simply not knowing what I am doing, as I have never previously used `cmake`, and have not used C++ since 1993… but as I am stuck, I hope you will humor me, and perhaps this can result in slightly more explicit instructions in the future.

I am trying to build a program which uses both Link and mongoose, an open-source networking library. I was able to get the basic mongoose proof-of-concept code working with `cmake` and using C++ to compile (even though mongoose itself is a C library), but as soon as I try adding the include for Link, compilation fails.

Here is the content of my `CMakeLists.txt`:

```cmake
cmake_minimum_required(VERSION 3.0)
project(Carabiner)

set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)

set(mongoose_HEADERS
  mongoose.h
)

set(mongoose_SOURCES
  mongoose.c
)

add_executable(Carabiner
  ${mongoose_HEADERS}
  ${mongoose_SOURCES}
  carabiner.cpp
)

include(~/git/link/AbletonLinkConfig.cmake)
target_link_libraries(Carabiner Ableton::Link)
```

And when I `mkdir build`, `cd build`, and `cmake ..`, that seems to work just fine. And I can compile (using `cmake --build .`) and run my source code (below) just fine, *as long as I omit the line which tries to include the Link header*. But as soon as that line is present, compilation fails with the errors shown below the source. This is on Mac OS X 10.11.6. (And when I follow the directions to create the Link example programs within the Link repository itself, that also works fine. The problem only occurs when I try to use Link from my own project.) Thanks in advance for any help you might be able to offer.

Here is `carabiner.cpp`:

```c++
extern "C" {
  #include "mongoose.h"
}
#include <ableton/Link.hpp>

static void ev_handler(struct mg_connection *nc, int ev, void *p) {
  struct mbuf *io = &nc->recv_mbuf;
  (void) p;
  switch (ev) {
    case MG_EV_RECV:
      mg_send(nc, io->buf, io->len);  // Echo message back
      mbuf_remove(io, io->len);       // Discard message from recv buffer
      nc->flags |= MG_F_SEND_AND_CLOSE;
      break;
    default:
      break;
  }
}

int main(void) {
  struct mg_mgr mgr;
  const char *port1 = "udp://1234", *port2 = "udp://127.0.0.1:17000";

  mg_mgr_init(&mgr, NULL);
  mg_bind(&mgr, port1, ev_handler);
  mg_bind(&mgr, port2, ev_handler);

  printf("Starting echo mgr on ports %s, %s\n", port1, port2);
  for (;;) {
    mg_mgr_poll(&mgr, 1000);
  }
  mg_mgr_free(&mgr);

  return 0;
}
```

And here are the problems which are reported trying to include the Link header:

```
cmake --build .                                          19:44:58
Scanning dependencies of target Carabiner
[ 33%] Building C object CMakeFiles/Carabiner.dir/mongoose.c.o
[ 66%] Building CXX object CMakeFiles/Carabiner.dir/carabiner.cpp.o
In file included from /Users/jim/git/carabiner/carabiner.cpp:21:
In file included from /Users/jim/git/link/include/ableton/Link.hpp:25:
In file included from /Users/jim/git/link/include/ableton/platforms/Config.hpp:22:
In file included from /Users/jim/git/link/include/ableton/link/Controller.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/Service.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/PeerGateways.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/InterfaceScanner.hpp:22:
In file included from /Users/jim/git/link/include/ableton/platforms/asio/AsioWrapper.hpp:54:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio.hpp:19:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/associated_executor.hpp:21:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/system_executor.hpp:19:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/scheduler.hpp:20:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/error_code.hpp:185:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/impl/error_code.ipp:29:
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/socket_types.hpp:300:9: error: 
      unknown type name 'ip_mreq'
typedef ip_mreq in4_mreq_type;
        ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/socket_types.hpp:304:9: error: 
      unknown type name 'ipv6_mreq'
typedef ipv6_mreq in6_mreq_type;
        ^
In file included from /Users/jim/git/carabiner/carabiner.cpp:21:
In file included from /Users/jim/git/link/include/ableton/Link.hpp:25:
In file included from /Users/jim/git/link/include/ableton/platforms/Config.hpp:22:
In file included from /Users/jim/git/link/include/ableton/link/Controller.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/Service.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/PeerGateways.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/InterfaceScanner.hpp:22:
In file included from /Users/jim/git/link/include/ableton/platforms/asio/AsioWrapper.hpp:54:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio.hpp:19:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/associated_executor.hpp:21:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/system_executor.hpp:19:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/scheduler.hpp:21:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/execution_context.hpp:406:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/impl/execution_context.hpp:20:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/service_registry.hpp:20:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/mutex.hpp:25:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/posix_mutex.hpp:71:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/impl/posix_mutex.ipp:24:
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/error.hpp:169:33: error: 
      use of undeclared identifier 'ESHUTDOWN'
  shut_down = ASIO_SOCKET_ERROR(ESHUTDOWN),
                                ^
In file included from /Users/jim/git/carabiner/carabiner.cpp:21:
In file included from /Users/jim/git/link/include/ableton/Link.hpp:25:
In file included from /Users/jim/git/link/include/ableton/platforms/Config.hpp:22:
In file included from /Users/jim/git/link/include/ableton/link/Controller.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/Service.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/PeerGateways.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/InterfaceScanner.hpp:22:
In file included from /Users/jim/git/link/include/ableton/platforms/asio/AsioWrapper.hpp:54:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio.hpp:21:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/basic_datagram_socket.hpp:21:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/datagram_socket_service.hpp:30:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/reactive_socket_service.hpp:30:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/reactive_socket_accept_op.hpp:24:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/socket_holder.hpp:20:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/socket_ops.hpp:333:
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/impl/socket_ops.ipp:120:21: error: 
      use of undeclared identifier 'SO_NOSIGPIPE'
        SOL_SOCKET, SO_NOSIGPIPE, &optval, sizeof(optval)), ec);
                    ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/impl/socket_ops.ipp:1405:21: error: 
      use of undeclared identifier 'SO_NOSIGPIPE'
        SOL_SOCKET, SO_NOSIGPIPE, &optval, sizeof(optval)), ec);
                    ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/impl/socket_ops.ipp:1498:23: error: 
      use of undeclared identifier 'SO_REUSEPORT'
          SOL_SOCKET, SO_REUSEPORT, optval, optlen);
                      ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/impl/socket_ops.ipp:2415:37: error: 
      no member named 'getipnodebyaddr' in the global namespace
  hostent* retval = error_wrapper(::getipnodebyaddr(
                                  ~~^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/impl/socket_ops.ipp:2470:37: error: 
      no member named 'getipnodebyname' in the global namespace
  hostent* retval = error_wrapper(::getipnodebyname(
                                  ~~^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/impl/socket_ops.ipp:2499:7: error: 
      no type named 'freehostent' in the global namespace
    ::freehostent(h);
    ~~^
In file included from /Users/jim/git/carabiner/carabiner.cpp:21:
In file included from /Users/jim/git/link/include/ableton/Link.hpp:25:
In file included from /Users/jim/git/link/include/ableton/platforms/Config.hpp:22:
In file included from /Users/jim/git/link/include/ableton/link/Controller.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/Service.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/PeerGateways.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/InterfaceScanner.hpp:22:
In file included from /Users/jim/git/link/include/ableton/platforms/asio/AsioWrapper.hpp:54:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio.hpp:78:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/ip/basic_resolver.hpp:28:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/ip/resolver_service.hpp:29:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolver_service.hpp:25:
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolve_endpoint_op.hpp:70:22: error: 
      use of undeclared identifier 'NI_MAXHOST'
      char host_name[NI_MAXHOST];
                     ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolve_endpoint_op.hpp:71:25: error: 
      use of undeclared identifier 'NI_MAXSERV'
      char service_name[NI_MAXSERV];
                        ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolve_endpoint_op.hpp:73:43: error: 
      use of undeclared identifier 'NI_MAXHOST'
          o->endpoint_.size(), host_name, NI_MAXHOST, service_name, NI_MAXSERV,
                                          ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolve_endpoint_op.hpp:73:69: error: 
      use of undeclared identifier 'NI_MAXSERV'
          o->endpoint_.size(), host_name, NI_MAXHOST, service_name, NI_MAXSERV,
                                                                    ^
In file included from /Users/jim/git/carabiner/carabiner.cpp:21:
In file included from /Users/jim/git/link/include/ableton/Link.hpp:25:
In file included from /Users/jim/git/link/include/ableton/platforms/Config.hpp:22:
In file included from /Users/jim/git/link/include/ableton/link/Controller.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/Service.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/PeerGateways.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/InterfaceScanner.hpp:22:
In file included from /Users/jim/git/link/include/ableton/platforms/asio/AsioWrapper.hpp:54:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio.hpp:78:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/ip/basic_resolver.hpp:28:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/ip/resolver_service.hpp:29:
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolver_service.hpp:93:20: error: 
      use of undeclared identifier 'NI_MAXHOST'
    char host_name[NI_MAXHOST];
                   ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolver_service.hpp:94:23: error: 
      use of undeclared identifier 'NI_MAXSERV'
    char service_name[NI_MAXSERV];
                      ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolver_service.hpp:96:20: error: 
      use of undeclared identifier 'NI_MAXHOST'
        host_name, NI_MAXHOST, service_name, NI_MAXSERV,
                   ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/resolver_service.hpp:96:46: error: 
      use of undeclared identifier 'NI_MAXSERV'
        host_name, NI_MAXHOST, service_name, NI_MAXSERV,
                                             ^
In file included from /Users/jim/git/carabiner/carabiner.cpp:21:
In file included from /Users/jim/git/link/include/ableton/Link.hpp:25:
In file included from /Users/jim/git/link/include/ableton/platforms/Config.hpp:22:
In file included from /Users/jim/git/link/include/ableton/link/Controller.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/Service.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/PeerGateways.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/InterfaceScanner.hpp:22:
In file included from /Users/jim/git/link/include/ableton/platforms/asio/AsioWrapper.hpp:54:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio.hpp:83:
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/ip/icmp.hpp:58:17: error: 
      use of undeclared identifier 'IPPROTO_ICMPV6'
    return icmp(ASIO_OS_DEF(IPPROTO_ICMPV6),
                ^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/socket_types.hpp:318:25: note: 
      expanded from macro 'ASIO_OS_DEF'
# define ASIO_OS_DEF(c) ASIO_OS_DEF_##c
                        ^
<scratch space>:61:1: note: expanded from here
ASIO_OS_DEF_IPPROTO_ICMPV6
^
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/detail/socket_types.hpp:331:37: note: 
      expanded from macro 'ASIO_OS_DEF_IPPROTO_ICMPV6'
# define ASIO_OS_DEF_IPPROTO_ICMPV6 IPPROTO_ICMPV6
                                    ^
In file included from /Users/jim/git/carabiner/carabiner.cpp:21:
In file included from /Users/jim/git/link/include/ableton/Link.hpp:25:
In file included from /Users/jim/git/link/include/ableton/platforms/Config.hpp:22:
In file included from /Users/jim/git/link/include/ableton/link/Controller.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/Service.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/PeerGateways.hpp:22:
In file included from /Users/jim/git/link/include/ableton/discovery/InterfaceScanner.hpp:22:
In file included from /Users/jim/git/link/include/ableton/platforms/asio/AsioWrapper.hpp:54:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio.hpp:84:
In file included from /Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/ip/multicast.hpp:20:
/Users/jim/git/link/cmake_include/../modules/asio-standalone/asio/include/asio/ip/detail/socket_option.hpp:403:25: error: 
      member reference base type 'asio::detail::in6_mreq_type' (aka 'int') is
      not a structure or union
      memcpy(ipv6_value_.ipv6mr_multiaddr.s6_addr, bytes.data(), 16);
             ~~~~~~~~~~~^~~~~~~~~~~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [CMakeFiles/Carabiner.dir/carabiner.cpp.o] Error 1
make[1]: *** [CMakeFiles/Carabiner.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by nre-ableton** [comment]
Do you happen to have a repo with the project on GitHub somewhere? Otherwise, just by looking at your `CMakeLists.txt` file, I would speculate that the line with `include(~/git/link/AbletonLinkConfig.cmake)
` should appear before the `add_executable` statement.

If this fixes the problem, we will update the documentation accordingly.

Also, if you are unfamiliar with CMake, it's also pretty easy to build a Link-based project with another build system of your choice. Instructions for doing so are in the README, which I presume you saw since your `CMakeLists.txt` follows the CMake instructions there.

**Comment by brunchboy** [comment]
Thanks for the suggestion, but it did not seem to help. In retrospect, should have created a GitHub project before posting this; I had been hoping to get a bit further along in this project before publishing anything, but running into a snag like this should have changed my plans.

I tried following the README instructions for non-CMake builds first, and ran into the same errors, which led me to believe that there was some system-specific configuration required, and that CMake might help me with it. That, plus the thought that CMake would make it easier for people to build this project on non-Mac platforms sent me down the route of trying to learn enough CMake to get it building.

I have published the repo as https://github.com/brunchboy/carabiner

In order to test it you'll probably need to edit the `AbletonLinkConfig.cmake` [include line](https://github.com/brunchboy/carabiner/blob/master/CMakeLists.txt#L4) to point to its correct location on your system. (As an aside, do you have a recommended way to make that easily configurable for people without actually embedding a checkout of Link in my own project? I’m more used to Java dependency management environments where I reference an artifact by its permanent release identifier.)

To recap, the errors above are occurring on OS X El Capitan 10.11.6, with Xcode 8.0 (8A218a), Homebrew 1.0.8, and cmake version 3.6.2. Thanks so much for helping look at this, and please let me know if there is any other information that would be of use.

**Comment by fgo-ableton** [comment]
It seem some defines in mongoose get in the way of asio. Try including `mongoose.h` after including `Link.hpp`.
In terms of dependency management I'd recommend adding ling as a git submodule to your project.

**Comment by brunchboy** [comment]
Thank you for identifying this issue. Yes, it appears to be a problem with polluting the global namespace, which reminds me of how nice it’s been to be working in a hierarchical environment like Java for the past many years!

This has gotten me un-stuck, and if I need to, I can try to keep the Mongoose-dependent files separate from those which depend on Link.

I was thinking the submodule approach might be how to go as well, although that means people will be dealing with three layers of submodules, since Link already does that with `asio-standalone` and `catch`. Hopefully the recursive submodule initialization process will traverse all of that!

**Comment by brunchboy** [comment]
I just wanted to thank you all again for your help. I have learned enough cmake to find it a very useful tool, and have been successful in creating my Link-based project which includes it as a Git submodule, and enables me to synchronize a Link session to the external clock established by a Pioneer Pro DJ Link network, cleanly tracking the BPM and aligning at the level of individual beats or entire bars.

Link has proved a delightfully designed protocol and I am excited to finish up and release this new feature, and then move on to enabling my lighting control software to synchronize its own timeline with a Link session, which will be an even easier use of the bridge I have created.
